packages/cli/src/services/FileCommandLoader.ts

```diff
@@ -0,0 +1,215 @@
+import toml from '@iarna/toml';
+import {
+  Config,
+  getProjectCommandsDir,
+  getUserCommandsDir,
+} from '@google/gemini-cli-core';
+import { ICommandLoader } from './types.js';
+import { CommandKind, SlashCommand } from '../ui/commands/types.js';
+
+/**
+ * Defines the structure of a valid command definition within a .toml file.
+ * This is the raw shape of the data after being parsed from TOML.
+ */
+interface TomlCommandDef {
+  prompt: string;
+  description?: string;
+}
+
+/**
+ * A type guard to safely validate that a parsed TOML object conforms to our
+ * required command definition structure.
+ *
+ * @param obj The unknown object parsed from a TOML file.
+ * @returns True if the object is a valid TomlCommandDef, false otherwise.
+ */
+function isTomlCommandDef(obj: unknown): obj is TomlCommandDef {
+  // Basic check for an object-like structure.
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+
+  // Check that 'prompt' is a non-empty string.
+  const hasPrompt =
+    'prompt' in obj && typeof (obj as TomlCommandDef).prompt === 'string';
+
+  // Check that 'description', if it exists, is a string.
+  const hasValidDescription =
+    !('description' in obj) ||
+    typeof (obj as TomlCommandDef).description === 'string';
+
+  return hasPrompt && hasValidDescription;
+}
+
+function isErrnoException(error: unknown): error is NodeJS.ErrnoException {
+  return error instanceof Error && 'code' in error;
+}
+
+/**
+ * Discovers and loads custom slash commands from .toml files in both the
+ * user's global config directory and the current project's directory.
+ *
+ * This loader is responsible for:
+ * - Recursively scanning command directories.
+ * - Parsing and validating TOML files.
+ * - Adapting valid definitions into executable SlashCommand objects.
+ * - Handling file system errors and malformed files gracefully.
+ */
+export class FileCommandLoader implements ICommandLoader {
+  private readonly projectRoot: string;
+
+  constructor(private readonly config: Config | null) {
+    this.projectRoot = config?.getProjectRoot() || process.cwd();
+  }
+
+  /**
+   * Loads all commands, applying the precedence rule where project-level
+   * commands override user-level commands with the same name.
+   * @param signal An AbortSignal to cancel the loading process.
+   * @returns A promise that resolves to an array of loaded SlashCommands.
+   */
+  async loadCommands(signal?: AbortSignal): Promise<SlashCommand[]> {
+    const userDir = getUserCommandsDir();
+    const projectDir = getProjectCommandsDir(this.projectRoot);
+
+    const [userCommands, projectCommands] = await Promise.all([
+      this.loadCommandsFromDir(userDir, userDir, signal),
+      this.loadCommandsFromDir(projectDir, projectDir, signal),
+    ]);
+
+    const commandMap = new Map<string, SlashCommand>();
+    for (const cmd of userCommands) {
+      commandMap.set(cmd.name, cmd);
+    }
+    for (const cmd of projectCommands) {
```

Comment:
Should add a comment here around the sequencing being very intentional for overrides